### ISSUE TITLE
I can not found Broadcast message to a connectionId

### ISSUE THREAD

**Issue by majidbigdeli** [kataras/neffos#1]
How Sending message to specific user with  neffos ? 
I can not found Broadcast message to a connectionId

**Comment by kataras** [comment]
Hello @majidbigdeli,

First of all I am very glad that you are posting questions while this library is still (on the final steps of) work-in-progress!

It's easy to do something like this, we just have to add a `"To"` field in the `Message` and check its value inside the connected namespaces (and rooms) on the `Server#Broadcast` but to complete a feature like this the "To" connection should be able to see who was the original **client** sender "From" (because you may want to print its name before the message's body). That was one of my internal questions I had when designing the library, if the library should or not to expose the connection ID (the original client publisher that server-side connection catch and calls its `()Server.Broadcast(Message{})`) of the message to the subscribers (to other clients) or not and back then I decided that it's better to not do it for security reasons (the `Server#IDGenerator` can be used to set connections' IDs based on a custom algorithm, a developer may set this ID to a private database user's row's ID, so we should **not expose the sender ID by a "From" field**).

At the other hand, the `"From"` can be solved by the end-developer him/herself by using a custom object and set the `Message.Body` to the object's serialized form(i.e JSON), and deserialize it on read. An example is also available for `encoding/json` and even protobufs, check [_examples/basic](https://github.com/kataras/neffos/blob/master/_examples/basic/main.go) and [_examples/protobuf](https://github.com/kataras/neffos/tree/master/_examples/protobuf).

As you probably already know from the Go and JavaScript examples,
so far the messages from `Server#Broadcast` are sent to all connections that are connected to the Message's specific Namespace (and room if `Message { Room: "is_not_empty"}`) like `NSConn,Room#Emit(...) method` but loops through all server's active connections.

**However**, we can have an optional `Message.To string` field among the `Namespace` and `Room` fields to decide if an active connection's Emit should be called **AND** a `From` field can be sent **IF** the given `To` field is not empty and first parameter of the `Server#Broadcast` is not nil. Sounds great?

**Comment by kataras** [comment]
Hello @majidbigdeli,

First of all I am very glad that you are posting questions while this library is still (on the final steps of) work-in-progress!

It's easy to do something like this, we just have to add a `"To"` field in the `Message` and check its value inside the connected namespaces (and rooms) on the `Server#Broadcast` but to complete a feature like this the "To" connection should be able to see who was the original **client** sender "From" (because you may want to print its name before the message's body). That was one of my internal questions I had when designing the library, if the library should or not to expose the connection ID (the original client publisher that server-side connection catch and calls its `()Server.Broadcast(Message{})`) of the message to the subscribers (to other clients) or not and back then I decided that it's better to not do it for security reasons (the `Server#IDGenerator` can be used to set connections' IDs based on a custom algorithm, a developer may set this ID to a private database user's row's ID, so we should **not expose the sender ID by a "From" field**).

At the other hand, the `"From"` can be solved by the end-developer him/herself by using a custom object and set the `Message.Body` to the object's serialized form(i.e JSON), and deserialize it on read. An example is also available for `encoding/json` and even protobufs, check [_examples/basic](https://github.com/kataras/neffos/blob/master/_examples/basic/main.go) and [_examples/protobuf](https://github.com/kataras/neffos/tree/master/_examples/protobuf).

As you probably already know from the Go and JavaScript examples,
so far the messages from `Server#Broadcast` are sent to all connections that are connected to the Message's specific Namespace (and room if `Message { Room: "is_not_empty"}`) like `NSConn,Room#Emit(...) method` but to all server's active connections (it does not use loop because it would require a lock to add/remove active connections, instead I used a custom `sync.Cond` therefore new connections are still able to connect and old can disconnect even while something is broadcasting).

**However**, we can have an optional `Message.To string` field among the `Namespace` and `Room` fields to decide if an active connection's Emit should be called **AND** a `From` field can be sent **IF** the given `To` field is not empty and first parameter of the `Server#Broadcast` is not nil. 

Sounds good to you?

**Comment by kataras** [comment]
It's done @majidbigdeli, just fill the `Message.To` field.

```go
server.Broadcast(c, Message{ To: "other connecton id", Namespace: "...", Room: "...", Body: ....})
```

I decided to not expose the `from` field from the Message at any case, end-developer can send and receive any custom data so they are able to fill if really required, as explained above.

**Comment by majidbigdeli** [comment]
Hello my friend @kataras .
I read all your commits every day. (Because you are a great programmer) 
I work In the Asterisk Manager Interface (AMI) project,
I need to work with the web socket.
I used socket.io before. But after I looked at the neffos project carefully.
I decided to use neffos instead of socket.io . neffos project is very clean project.

Finally, I think you made the right decision.
In my opinion, If someone needs to know who is the original client sender of the message, they can use the custom data.
I agree with this .(can be solved by the end-developer him/herself by using a custom object and set the Message.Body to the object's serialized form(i.e JSON), and deserialize it on read.)

@kataras   Thanks for implementing Message.To. ❤️❤️

@kataras  Thanks for implementing neffos and neffos.js . ❤️❤️

**Comment by kataras** [comment]
This post of yours, my friend @majidbigdeli,  is one of the most inspirational messages I have ever received for my work, I am really thankful!

Yes, I had a lot of issues working with socket.io at my past career as well. I think a library based on raw websockets is the best option nowdays, note that socket.io has a lot of steps to decide what subprotocol to use for communication(this is the top reason makes that uncompatible with nodejs `ws` and browser's websockets by-design) and a lot of code which is not even readable for most of the developers over the years. This is one of the main reasons I decided to design and code a decent alternative from zero, hopefully this will be more visible by next months when Iris v11.2 will be released as well ❤️

**Comment by lauretagabriel** [comment]
Hi @kataras, may I know the estimated date where this project will be fully integrated and stable with the Iris framework? Sorry to sound demanding but really looking forward to use this together with the Iris framework. Thank you!!

**Comment by kataras** [comment]
Hello @lauretagabriel, I feel you, it's a good question. Expect it until the end of the next week, maybe a bit sooner.

**Comment by kataras** [comment]
@lauretagabriel @majidbigdeli can I ask you to give some time in order to read and give me feedback about the new [neffos wiki](https://github.com/kataras/neffos/wiki) Getting Started page, is it understable enough for newcomers and for you? Generally offer some feedback on how to process with it if you want, I want to minimize the README and push a complete guide inside the project's wiki which can be converted to a special book on the future as well. Thank you a lot guys.

**Comment by majidbigdeli** [comment]
@kataras  Hello .
I read this page and I will give you feedback.
Thankful.

**Comment by kataras** [comment]
@majidbigdeli Thanks a lot!

About JMeter, we can use JMeter with neffos only with the `native-only messages` because as you understand the JMeter and its plugins are using the websocket API but neffos has its own way to verify a client and to support namespaces, rooms and etc, so we can't use them directly with JMeter because it will stack on the "ack" part of the neffos, they will not pass unless the same neffos websocket client steps are produced. However, if you are an experienced JMeter user you are more fit to do that and share the results with the rest of us.

I'll add a wiki page about native messages, although Iris v11.2 has its own example already, but it looks like this:

![img43](https://user-images.githubusercontent.com/22900943/60385113-6f02d180-9a8e-11e9-9206-35e1a80c0ea5.png)

**Comment by kataras** [comment]
^ https://github.com/kataras/neffos/wiki/Native-messages

**Comment by majidbigdeli** [comment]
> ^ https://github.com/kataras/neffos/wiki/Native-messages

Thank you for Example.

Yes.  I just tested to open the connection.
Πραγματικά εκτιμώ όλη τη σκληρή δουλειά που έκανες για να με βοηθήσεις. ❤️ ❤️

**Comment by majidbigdeli** [comment]
@kataras 
When I disconnect and reconnect the network, I get error.
neffos.min.js?v980406:1 Fetch API cannot load wss://wss.ami.dev.iwin.local:3811/echo. URL scheme must be "http" or "https" for CORS request.

![image](https://user-images.githubusercontent.com/26275762/60402113-45d26600-9ba0-11e9-9dec-316d65919922.png)

**Comment by kataras** [comment]
Thanks for your nice words and the greek translation too!!! I really appreciate it!!! I am so lucky having you around too! I answered you at the neffos.js project: https://github.com/kataras/neffos.js/issues/2#issuecomment-507076491

**Comment by majidbigdeli** [comment]
@kataras 

```
server := neffos.New(gobwas.DefaultUpgrader, serverAndClientEvents)
	server.IDGenerator = func(w http.ResponseWriter, r *http.Request) string {
		if userID := r.Header.Get("X-Username"); userID != "" {
			return userID
		}

		return neffos.DefaultIDGenerator(w, r)
	}

```

I want to use this form.
with  iris.Context. it is not issue. I love to use iris.Context 

```
server := neffos.New(gobwas.DefaultUpgrader, serverAndClientEvents)
	server.IDGenerator = func(ctx iris.Context) string {
		if userID := ctx.GetHeader("X-Username"); userID != "" {
			return userID
		}
		return neffos.DefaultIDGenerator(w, r)
	}
```

**Comment by kataras** [comment]
Hey @majidbigdeli, 

I don't want to depend on Iris for the neffos package because it will make its download heavier, instead I added this feature on the Iris side itself. Using the `websocket.Handler` which now accepts an **optional second argument of func(ctx iris.Context) string**:

```go
import "github.com/kataras/iris/websocket"

[...]
irisIDGenerator := func(ctx iris.Context) string {
  return ctx.GetHeader("X-Username")
}
app.Get("/my_endpoint", websocket.Handler(websocketNeffosServer, irisIDGenerator))
```

Wiki: https://github.com/kataras/neffos/wiki/Getting-started#run-the-server

**Comment by majidbigdeli** [comment]
@kataras 

> I don't want to depend on Iris for the neffos package because it will make its download heavier, instead I added this feature on the Iris side itself. Using the `websocket.Handler` which now accepts an **optional second argument of func(ctx iris.Context) string**:

**Exactly**. it's true .
You're really a great developer.
I really enjoy your ideas and learn.

**Comment by kataras** [comment]
**Thank YOU** @majidbigdeli, you are free to give me feedback about the wikis as well, you can give some language improvements indeed or if you found any spelling mistakes that I missed please let me know, at the same time I am collecting and improving the iris documentation to one single place as well

**Comment by majidbigdeli** [comment]
@kataras  
Yes.I read all the wiki and pdf pages. Thanks for the pdf That you sent me.
The wiki pages were well-ordered and I enjoyed it.
For pdf, I think the fonts should be more appropriate.
I'll read the wiki again.
**Your work is commendable.** 🙏🙏🙏

**Comment by majidbigdeli** [comment]
@kataras

How Wait for response from client?

 server   *neffos.Server
server does not have the ask().

For example, this form
```
server.ask(ctx,neffos.Message{
		To:        extenstionString,
		Namespace: namespace,
		Event:     showForm,
		Body:      output,})

```

**Comment by kataras** [comment]
@majidbigdeli Both server and client connections shares the same methods, for `Conn` you have to use the   [Conn.Ask](https://github.com/kataras/neffos/blob/26005b6ec445554c1155d1373a3605e23ad16c55/conn.go#L742) and inside events, or `nsConn := conn.Connect(...)` you can also use the [NSConn.Ask](https://github.com/kataras/neffos/blob/26005b6ec445554c1155d1373a3605e23ad16c55/conn_namespace.go#L60).


**method 1**
```go
// [server := neffos.New(..., events)]
// nsConn := server.Connect(ctx, "namespace") or
// let the client connect and use `WaitConnect(ctx, "namespace")` to wait for it

response, err := nsConn.Ask(ctx, neffos.Message{...})
```
**method 2**
```go
var events = neffos.Events {
   // if you want to wait for a message at the namespace connected.
    neffos.OnNamespaceConnected( c *neffos.NSConn, msg neffos.Message) error {
         response, err := c.Ask(ctx, neffos.Message{...})
    },
    // if you want to wait for a message inside an event.
    "myEvent": func(c *neffos.NSConn, msg neffos.Message) error {
            response, err := c.Ask(ctx, neffos.Message{...})
   },
}
```

**Comment by majidbigdeli** [comment]
@kataras  .Ye It is true.

I can not use method 2 because i brodcast messages From inside Api

```
func broadcastHandler(w http.ResponseWriter, r *http.Request) {

	var userMsg userMessage

	if r.Body == nil {
		http.Error(w, "Please send a request body", 400)
		return
	}
	err := json.NewDecoder(r.Body).Decode(&userMsg)
	if err != nil {
		http.Error(w, err.Error(), 400)
		return
	}

	output, err := json.Marshal(userMsg)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}

	extensionMessage := strconv.Itoa(userMsg.Extension)

	server.Broadcast(nil, neffos.Message{
		To:        extensionMessage,
		Namespace: namespace,
		Event:     showForm,
		Body:      output,
	})

	data.InsertLogForForm(userMsg.Extension, userMsg.Direction, 1, userMsg.CallID, userMsg.CallerNumber)

	w.WriteHeader(http.StatusOK)
}
```

**Comment by majidbigdeli** [comment]
@kataras 
also I said to my friend who is an English professor that he would read the wiki and tell me if it was a typo.

**Comment by majidbigdeli** [comment]
@kataras  for method 1  

[server := neffos.New(..., events)
 nsConn := server.Connect(ctx, "namespace") 

server does not have the Connect() 
.
yes . 
Conn.Ask  is exist  But I can not use

**Comment by kataras** [comment]
So you want to send something from a http request to a known upgraded websocket connection and block until response, you may want to check out the [Server.Do](https://github.com/kataras/neffos/blob/6031b3fb979c98d65080c0ce14fe94ad39539c3d/server.go#L383) method, example:

```go
server.Do(func(c *neffos.Conn){
   if c.ID() == extensionMessage {
      response, err := c.Ask(...)
      // [...]
   }
   
}, false)
```

Aboutthe `server.Connect`, yes you are right, sorry for my mistake, it couldn't be possible :P `conn.Connect` I meant, do it  on:
````go
server.OnConnect = (func(c *neffos.Conn) error{
        c.Connect("namespace") // to force-connect a client connection to a namespace.
})
````

**Comment by kataras** [comment]
@majidbigdeli Thanks!!! Let's wait for a review then.

Also, I think you will find interesting the [latest feature](
https://github.com/kataras/neffos/blob/6031b3fb979c98d65080c0ce14fe94ad39539c3d/_examples/redis/main.go#L16-L29) I've pushed yesterday as requested on #3

**Comment by majidbigdeli** [comment]
@kataras  I read [#3](https://github.com/kataras/neffos/issues/3) . Last night.
It is god Feature That I used in Signalr But Signalr does not suport redis clusters.
**It is very good feature** . I did not think it would be implemented in the zero version.
Thank You .

Then
Yes . I want to send something from a http request to a known upgraded websocket connection and block until response. thank you.

Yes . I can use server.Do

But I can not  use

```
server.OnConnect = func(c *neffos.Conn) error {
     nsConn :=  c.Connect(ctx, "namespace") // to force-connect a client connection to a namespace.
     // or let the client connect and use
     // nsConn := c.WaitConnect(ctx, "namespace") to wait for it.
     response, err := nsConn.Ask(ctx, neffos.Message{...})
}
```

**Comment by kataras** [comment]
Yes, `Do` is slow as documented but your other option is to cache the connections by yourself through a `map[string]*neffos.Conn` and retrieve by conn id manually, this would require to delete a connection on a `server.OnDisconnect` as well. Neffo does not keep a connection map like this because one or more connections may share the same ID and for other performance reasons.

**Comment by majidbigdeli** [comment]
@kataras 

I really excuse me. My English is not so good  .

I have cron that run every 1 second 
```
	c := cron.New()
	_ = c.AddFunc("@every 1s", func() {
		controller.NotificationHandler()
	})
	c.Start()
```

And 

NotificationHandler()

```
func NotificationHandler() {
	mutex.Lock()

	//listExtensionNumber , connectionID = extentionNumber
	var listExtensionNumber []model.IDTvp

	connections := Server.GetConnections()
	for c := range connections {
		var extentionNumber model.IDTvp
		connectionID, err := strconv.Atoi(c)
		if err != nil {
			fmt.Println(err)
		}
		extentionNumber.ID = connectionID
		listExtensionNumber = append(listExtensionNumber, extentionNumber)
	}

	//get list Notification from database with send list ExtensionNumber ;
	// Attention { list ExtensionNumber == list ConnectionID }
	listNotification, err := data.GetNotificationList(listExtensionNumber)

	if err != nil {
		fmt.Println(err)
		return
	}

	if listNotification != nil {
		if len(*listNotification) > 0 {
			for _, element := range *listNotification {
				//extensionNumber = my unique connection ID
				extensionNumber := strconv.Itoa(element.Number)
				Server.Broadcast(nil, neffos.Message{
					To:        extensionNumber,
					Namespace: variable.Agent,
					Event:     variable.OnNotification,
					Body:      neffos.Marshal(element),
				})
			}

		}
	}

	mutex.Unlock()

}

```
 Attention { Extension Number == Connection Id }

 I need to get a list of notifications with status = start  from the database every 1 second
And send any notifications to an user, and if user receive the notification , I  update  status of  notification in the database to send.

The problem is That if cron run again before update notification  status  ,a notification send again to user.

Because i use Server.Broadcast  I can not wait for response so cron run again 
```
Server.Broadcast(nil, neffos.Message{
					To:        extensionNumber,
					Namespace: variable.Agent,
					Event:     variable.OnNotification,
					Body:      neffos.Marshal(element),
				})
```

for  solution 1 neffos have
```
func (s *Server) GetConnections() map[string]*Conn {
	conns := make(map[string]*Conn)

	s.mu.RLock()
	for c := range s.connections {
		conns[c.ID()] = c
	}
	s.mu.RUnlock()

	return conns
}
```
i can use
`server.GetConnections()[extensionNumber].Ask(ctx, neffos.Message{...})`

**Comment by kataras** [comment]
@majidbigdeli I've just pushed a `Server.Ask(context, message)` method, can you test it out and tell me the results?. If "msg.To" is filled then it sends the request to a specific connection otherwise the first responder's reply will be returned instead.

**Comment by majidbigdeli** [comment]
@kataras  . 
I dont see reply() method in neffos.js .

**Comment by majidbigdeli** [comment]
@kataras 
i get online connections with 
```
	connections := Server.GetConnections()
	for c := range connections {
		var extentionNumber model.IDTvp
		connectionID, err := strconv.Atoi(c)
		if err != nil {
			fmt.Println(err)
		}
		extentionNumber.ID = connectionID
		listExtensionNumber = append(listExtensionNumber, extentionNumber)
	}
```
Because the server is blocked.
What is your offer to get online connections?

**Comment by kataras** [comment]
You don't need this @majidbigdeli, `Reply(body)` is just a helper for `nsConn.Emit(incomingMsg.Event, body)`

**Comment by majidbigdeli** [comment]
> helper for

thank you

**Comment by kataras** [comment]
> i get online connections with Server.GetConnections(), what is your offer to get online connections? because it call every 1s.

Move your whole logic PER CONN inside the  `server.OnConnect`, it's the best and most correct approach.

**Comment by majidbigdeli** [comment]
@kataras 
I tested `server.ask()` it's work. thank you .

**Comment by kataras** [comment]
OK nice @majidbigdeli, check my updated answer above and tell me if that helped you with the cron thing, otherwise I will push an example :P  So it's time to think a way to make this `Server.Ask` feature accessible when users use a `StackExchange` like redis or nats for scaling-out.

**Comment by majidbigdeli** [comment]
Do you suggest that for each connection makes a cron job ?
I have a lot of connections . 
As a result, a large number of requests go to the database at a time.

please see  [https://github.com/majidbigdeli/neffosAmiClient](https://github.com/majidbigdeli/neffosAmiClient)

i send list of connectionId(extentionNumber) to database and  i get list of notification

**Comment by majidbigdeli** [comment]
> OK nice @majidbigdeli, check my updated answer above and tell me if that helped you with the cron thing, otherwise I will push an example :P So it's time to think a way to make this `Server.Ask` feature accessible when users use a `StackExchange` like redis or nats for scaling-out.

yes .  Thank you if you push an example.

**Comment by kataras** [comment]
> Do you suggest that for each connection makes a cron job ?

No, I suggested to return a wrapper for this but now I see the whole code you use, and I understand some things, please answer the above so I can provide more help:

- Each database entry ID is linked with a websocket connection ID, this link is done on server.IDGenerator
- You have other database table with IDs subcollection in a database that are allowed to get notifications, not every database ID is allowed to get notified
- and all these can change at runtime

So that you need is a `Broadcast` to more than one `To` fields that depend on a database runtime call.

My recommendation is very simple and efficient:

Add a `message.ToFunc(c *neffos.Conn) bool` which will perform custom checks when a connection is allowed to send the broadcasted message from a `server.Broadcast` call. 

In that `ToFunc` you will make a select database call filtered by the `c.ID()` and return `true` if that matches the db criteria to get notified.

Sounds good?

**Comment by majidbigdeli** [comment]
every web socket connection Id is a Extension . Extension is User Id
```
	server.IDGenerator = func(w http.ResponseWriter, r *http.Request) string {
		if extension := r.Header.Get("Extension"); extension != "" {
			return extension
		}
		return neffos.DefaultIDGenerator(w, r)
	}

	server.OnUpgradeError = func(err error) {
		log.Printf("ERROR: %v", err)
	}
```
I cant not use Room or neffos.Message without To
Because  Each user has its own notifications (messages).

I get listExtention with 

```
	var listExtensionNumber []model.IDTvp

	connections := Server.GetConnections()
	for c := range connections {
		var extentionNumber model.IDTvp
		connectionID, err := strconv.Atoi(c)
		if err != nil {
			fmt.Println(err)
		}
		extentionNumber.ID = connectionID
		listExtensionNumber = append(listExtensionNumber, extentionNumber)
	}
```
then send listExtensionNumber to store procedure with
`
data.GetNotificationList(listExtensionNumber)`


method **GetNotificationList(listExtentions []model.IDTvp)** call one sp from database
```
func GetNotificationList(listExtentions []model.IDTvp) (*[]model.Notification, error) {
	var poError int32
	var poSTEP int32

	if len(listExtentions) > 0 {
		tvpType := mssql.TVP{
			TypeName: "typ.BigIntIDType",
			Value:    listExtentions,
		}

		var d []model.Notification
		err := dbData.Select(&d, "app.uspGetNotificationList1",
			sql.Named("PI_Extensions", tvpType),
			sql.Named("PO_Error", sql.Out{Dest: &poError}),
			sql.Named("PO_STEP", sql.Out{Dest: &poSTEP}),
		)
		return &d, err
	}
	return nil, nil
}
```
I'm just going to get the list of online User (connection Id) and send out their own messages for each one.  every  user  has its own notifications (messages).

Notification struct is

```
type Notification struct {
	NotificationID int    `json:"NotificationId" db:"NotificationId"`
	UserID         int    `json:"UserId" db:"UserId"`
	Type           int    `json:"Type" db:"Type"`
	Status         int    `json:"Status" db:"Status"`
	MessageType    int    `json:"MessageType" db:"MessageType"`
	Data           string `json:"Data" db:"Data"`
	Message        string `json:"Message" db:"Message"`
	ExtensionID    int    `json:"ExtensionID" db:"ExtensionID"`
	Number         int    `json:"Number" db:"Number"`
}

```


that Number in Notification struct (Notification.Number) is equal with Extension (connection Id)

I get list of Notification struct from db []Notification So 

```
//listNotification is []Notification
listNotification, err := data.GetNotificationList(listExtensionNumber)

	if err != nil {
		fmt.Println(err)
		return
	}

	if listNotification != nil {
		if len(*listNotification) > 0 {
			for _, element := range *listNotification {
                                 //element.Number is extension (connection Id)  In the form of int
				//extensionNumber = is extension (connection Id)  in the from string
                                // convert int to string
				extensionNumber := strconv.Itoa(element.Number)
				Server.Broadcast(nil, neffos.Message{
					To:        extensionNumber,
					Namespace: variable.Agent,
					Event:     variable.OnNotification,
					Body:      neffos.Marshal(element),
				})
			}

		}
	}
```

**Comment by majidbigdeli** [comment]
**I just need a List connection Id.**
I get List connection Id  with
```
var listConnectionId []string

	connections := Server.GetConnections()
	for c := range connections {
		listConnectionId = append(listConnectionId, c)
	}
```

**Comment by majidbigdeli** [comment]
> * Each database entry ID is linked with a websocket connection ID, this link is done on server.IDGenerator
Each **websocket connection ID** has multiple messages.
Each  **websocket connection ID** has different messages.

**Comment by majidbigdeli** [comment]
> However, such a feature like this couldn't be ever compatible with StackExchange across different neffos instances because of the func


This change is not required.

**Comment by majidbigdeli** [comment]
> **UPDATE** `Conn.Set` and `Conn.Get` methods added.


Thank you for [this](d1041e7d89d7975b16261b6f08e5fb952d742092)

**Comment by kataras** [comment]
> 
> 
> **I just need a List connection Id.**
> I get List connection Id with
> 
> ```
> var listConnectionId []string
> 
> 	connections := Server.GetConnections()
> 	for c := range connections {
> 		listConnectionId = append(listConnectionId, c)
> 	}
> ```

As I already mentioned neffos does not keep a collection like this because by-default there is no guarante that an ID is mapping only to one single connection and for other performance reasons.
So you have to do it by yourself, using channels or locks, the example below is a simple one using a slice with rw locks.

```go
var (
    listConnectionId []string
    listConnectionIdMutex sync.RWMutex
)
server.OnConnect = func(c *neffos.Conn) error {
    listConnectionIdMutex.Lock()
    listConnectionId = append(listConnectionId, c.ID())
    listConnectionIdMutex.Unlock()
}

server.OnDisconnect = func(c *neffos.Conn) {
    listConnectionIdMutex.RLock() // does not block the write locks.
    // find index to remove here
    indexToRemove := -1
    for i,n:=0, len(listConnectionId);i<n;i++ {
         if listConnectionId[i] == c.ID() {
            indexToRemove = i
            break
         }
    }
    listConnectionIdMutex.RUnlock()

    if indexToRemove != -1 {
        listConnectionIdMutex.Lock()
        if len(listConnectionId) == 1 {
              listConnectionId = listConnectionId[0:0]
        }else{
             listConnectionId  = 
             append(listConnectionId [:indexToRemove ], listConnectionId [indexToRemove +1:]...)
        } 
        listConnectionIdMutex.Unlock()
    }
}
```

However as I understood in your code, you don't want to send notifications to all websocket connections, right? Because if you just want to send to all you know that you can use the `server.Broadcast(.. msg)` without a `msg.To` field.
Instead you want to compare those with another list at your: `listNotification, err := data.GetNotificationList(listExtensionNumber)`, but I dont understand why you get a list of notification list IDs each single time? You can do a select statement and check if a specific ID is in the list, per conn and you don't need the listConnectionId at all.

I am preparing an example for you, it will take some time because I do 1000 things in the same time:p

**Comment by kataras** [comment]
@majidbigdeli `neffos.reply` is pushed on [neffos.js](https://github.com/kataras/neffos.js), works exactly the same as the Go's side.

**Comment by majidbigdeli** [comment]
> **UPDATE** I tried to follow the logic of your application and pushed an example at: https://github.com/kataras/neffos/blob/master/_examples/cronjob/main.go - that example shows various things that may help you in general, please tell me if that helped.

@kataras  . I do not have any brother but you are like a brother Thank You.
Yes, that's exactly what I wanted.

I'm going to send messages to all web socket connection Id . But i can not use neffos.message With out 
To  Because websocket connection id 1 have ( message id 1 , messageId 2 ,messageId 3) and websocket connection id 2 have (messageId 4 , messageId 5)  and websocket connection id 3 have (messageId 6 , messageId 8, messageId 12,)  .

websocket connection Id 1 should not see messageId 5 .

To determine messages for user  I send  list  websocket connection Id  to the database .

I just needed  list websocket connection id without server block .

your example is very good..

**Comment by kataras** [comment]
Thanks @majidbigdeli I will try to not let you down then!! I am very honored, I have two smaller brothers though :P

So that's exactly the example does, we will let it there, and if you want you can improve it, for other users that may have the same application requirements as yours, sounds good?

**Comment by kataras** [comment]
Thanks @majidbigdeli I will try to not let you down then!! I am very honored.

So that's exactly the example does, we will let it there, and if you want you can improve it, for other users that may have the same application requirements as yours, sounds good?

**Comment by majidbigdeli** [comment]
@kataras  Thank You .
Yes .  I'm going to implement an real world project with iris web socket version 11.2  and push it into iris. 
For example iris 11.2 with  Asterisk Manager Interface (AMI) . Thank you so much. 
I like to translate iris into Persian, but my English language is not good, but I will try.
@kataras  Thank you for this amazing support.

**Comment by majidbigdeli** [comment]
>  I don't want to depend on Iris for the neffos package because it will make its download heavier, instead I added this feature on the Iris side itself. Using the `websocket.Handler` which now accepts an **optional second argument of func(ctx iris.Context) string**:
> 
> ```go
> import "github.com/kataras/iris/websocket"
> 
> [...]
> irisIDGenerator := func(ctx iris.Context) string {
>   return ctx.GetHeader("X-Username")
> }
> app.Get("/my_endpoint", websocket.Handler(websocketNeffosServer, irisIDGenerator))
> ```
> 
> Wiki: https://github.com/kataras/neffos/wiki/Getting-started#run-the-server


@kataras 
ctx.GetHeader("X-Username")  Always return "" ;

**Comment by majidbigdeli** [comment]
@kataras  I have big issue with iris web socket  v11.2

when i use net/http with neffos and neffos.js in production with many request Everything is ok.
but when i use iris v11.2 websocket in production with many request is not good . I do not know why .
my work with iris 

```
app.Post("/broadcast", func(ctx iris.Context) {
		var userMsg dto.FormDto

		if err := ctx.ReadJSON(&userMsg); err != nil {
			ctx.StatusCode(iris.StatusBadRequest)
			return
		}
		userID, err := data.GetUserIDByExtentionNumber(userMsg.Extension)
		if err != nil {
			ctx.StatusCode(iris.StatusBadRequest)
			return
		}
		userNumber := strconv.Itoa(userID)

		output, err := json.Marshal(userMsg)

		if err != nil {
			ctx.StatusCode(iris.StatusBadRequest)
			return
		}
		websocketServer.Broadcast(nil, websocket.Message{
			To:        userNumber,
			Namespace: variable.Agent,
			Event:     variable.OnShowForm,
			Body:      output,
		})

		data.InsertLogForForm(userMsg.Extension, userMsg.Direction, 1, userMsg.CallID, userMsg.CallerNumber)

		ctx.StatusCode(http.StatusOK)

	})
```
my repo

[https://github.com/majidbigdeli/neffosAmiClient](https://github.com/majidbigdeli/neffosAmiClient
)

branch master is net/http with neffos and branch iris is iris v11.2 web socket .
please help . I do not know why . in development is ok but in production no .

**Comment by majidbigdeli** [comment]
> **UPDATE** I tried to follow the logic of your application and pushed an example at: https://github.com/kataras/neffos/blob/master/_examples/cronjob/main.go - that example shows various things that may help you in general, please tell me if that helped.


@kataras  cron job example when multi tap is open when close one tap not work

**Comment by kataras** [comment]
You have right @majidbigdeli , this happens because the parsing of url params as headers (for browser side header settings) it's done after the Iris' id generator call.

Fixed now, it will be pushed on the [Iris v11.2.3](https://github.com/kataras/iris/pull/1316) and the neffos 0.0.9 version (I made a change here too which fixes this issue and other issue that an Iris ID generator was called before the handshake, which logically is wrong).
